### PR TITLE
Added new doc for Schema

### PR DESCRIPTION
**Pull request by NotRobSheridan**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
New doc added to explain schema gen, when to generate schema, source control considerations, etc.
Edited links throughout the docs (Though I might have missed a few) 

#### Tests
Local Improbadoc build

#### Primary reviewers
Laura (though she has already pontificated on the doc), Chris, Giray.

### CODE REVIEW

**Comment by NotRobSheridan** [comment]
Note for @ElleEss - I've only gone through and edited links on a bunch of pages, I haven't touched them otherwise. Main things to review are the schema page and the glossary entry.

File: SpatialGDK/Documentation/content/get-started/tutorial.md [hunk 1]
```diff
@@ -44,7 +44,7 @@ In this step, you're going to build the Unreal GDK's dependencies.
 1. In the same directory, open **ThirdPersonShooter.sln** with Visual Studio.
 1. In the Solution Explorer window, right-click on **ThirdPersonShooter** and select **Build**.
 1. Open **ThirdPersonShooter.uproject** in the Unreal Editor.
-1. In the [GDK toolbar (SpatialOS documentation)]({{urlRoot}}/content/toolbars), click [`Schema` (SpatialOS documentation)]({{urlRoot}}/content/glossary#schema) to generate schema and then [`Snapshot` (SpatialOS documentation)]({{urlRoot}}/content/glossary#snapshot) to generate a snapshot.
+1. In the [GDK toolbar]({{urlRoot}}/content/toolbars), click [`Schema`]({{urlRoot}}/content/schema) to generate schema and then [`Snapshot`]({{urlRoot}}/content/glossary#snapshot) to generate a snapshot.
```
Comment (NotRobSheridan) [inline SpatialGDK/Documentation/content/get-started/tutorial.md]: Note for @ElleEss - This page is on my todo list. 
For this PR I only edited the link, I can see a bunch of things that need changing but I'm gonna sort all that in a separate PR later on. <3

**Comment by chrisfields0** [comment]
I reviewed this in the google doc we've been sharing. @girayimprobable is there anything else you'd like to add to this or are we clear to go ahead and have this approved?

File: SpatialGDK/Documentation/content/generating-a-snapshot.md [hunk 2]
```diff
@@ -47,7 +47,7 @@ If you want your snapshots to be exported to a different path you can specify th
 ## When to generate a snapshot
 You need to regenerate a snapshot when:
```
Comment (ElleEss) [inline SpatialGDK/Documentation/content/generating-a-snapshot.md]: Point of clarity - just realised you generate a snapshot _after_ you have done these, not during (as suggested by present continuous verb (ing) here).
```suggestion
You need to regenerate a snapshot when you have:
```

File: SpatialGDK/Documentation/content/generating-a-snapshot.md [hunk 3]
```diff
@@ -47,7 +47,7 @@ If you want your snapshots to be exported to a different path you can specify th
...
 
-* Generating [schema]({{urlRoot}}/content/glossary#schema) for a new class.
+* Generating [schema]({{urlRoot}}/content/schema) for a new class.
```
Comment (ElleEss) [inline SpatialGDK/Documentation/content/generating-a-snapshot.md]:
```suggestion
* Generated [schema]({{urlRoot}}/content/schema) for a new class.
```

File: SpatialGDK/Documentation/content/glossary.md [hunk 4]
```diff
@@ -43,6 +43,10 @@ The GSM lists both the URL of the [Map (or Level - see Unreal documentation)](ht
 ### GSM
 Short for [Global State Manager](#global-state-manager).
 
+### SchemaDatabase
+
+The `SchemaDatabase.uasset` file contains information about uobjects and associated [schema (SpatialOS documentation)](https://docs.improbable.io/reference/13.6/shared/concepts/schema#schema) in your project. Information is automatically added to the SchemaDatabase by the GDK whenever you generate schema. You cannot manually edit this file. 
+
 ### Schema Generation
```
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/glossary.md]: Yes. (I didn't really look at this page holistically, just added the SchemaDatabase entry)
Fixed in latest commit.

File: SpatialGDK/Documentation/content/glossary.md [hunk 5]
```diff
@@ -43,6 +43,10 @@ The GSM lists both the URL of the [Map (or Level - see Unreal documentation)](ht
 ### GSM
 Short for [Global State Manager](#global-state-manager).
 
+### SchemaDatabase
+
+The `SchemaDatabase.uasset` file contains information about uobjects and associated [schema (SpatialOS documentation)](https://docs.improbable.io/reference/13.6/shared/concepts/schema#schema) in your project. Information is automatically added to the SchemaDatabase by the GDK whenever you generate schema. You cannot manually edit this file.
```
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/glossary.md]: I think it should be when you're referring to UObjects or the UObject class, but obvs not when you're typing out the file name. 
Changed. Thanks.

File: SpatialGDK/Documentation/content/schema.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+### When to generate schema
+
+You must generate schema:
+
+- When you add or change any replicated properties that you want to deploy to SpatialOS.
+- When you create a new map. This ensures that you generate schema for any unique objects on this map. 
+- When you open a map and you haven’t generated schema for it yet. 
+
+You must also generate a snapshot after generating schema by selecting the **Snapshot** button on the GDK toolbar. For more information about snapshots, refer to the [snapshot documentation]({{urlRoot}}/content/generating-a-snapshot).
```
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Spoke to @m-samiec and this isn't true anymore, snapshot generation has changed a fair bit in the last 5 minutes or so. The whole snapshot paragraph has been removed.

File: SpatialGDK/Documentation/content/schema.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+As the GDK automatically generates all the schema you need, you do not have to write or edit schema manually when using the GDK.
+
+### When to generate schema
+
+You must generate schema:
+
+- When you add or change any replicated properties that you want to deploy to SpatialOS.
+- When you create a new map. This ensures that you generate schema for any unique objects on this map. 
+- When you open a map and you haven’t generated schema for it yet.
```
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: <3

File: SpatialGDK/Documentation/content/schema.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+- When you add or change any replicated properties that you want to deploy to SpatialOS.
+- When you create a new map. This ensures that you generate schema for any unique objects on this map. 
+- When you open a map and you haven’t generated schema for it yet. 
+
+You must also generate a snapshot after generating schema by selecting the **Snapshot** button on the GDK toolbar. For more information about snapshots, refer to the [snapshot documentation]({{urlRoot}}/content/generating-a-snapshot).
+
+The GDK only generates schema for objects currently loaded into memory. This means if your project uses [sublevels](<https://docs.unrealengine.com/en-us/Engine/Levels/LevelsWindow>), you’ll need to load them in addition to your map, before generating schema.
+
+Schema [component IDs](https://docs.improbable.io/reference/13.5/shared/schema/reference#ids) persist between runs of the schema generator. This means that as long as you have generated schema for a particular map at least once, and have not modified any classes or maps since you last generated schema, then you only need to generate a snapshot for the current level before you start a SpatialOS deployment.
```
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Due to the snapshot changes we no longer need this section so I've culled it. This info is implicit in the above "when to generate schema" section.

File: SpatialGDK/Documentation/content/schema.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+You must also generate a snapshot after generating schema by selecting the **Snapshot** button on the GDK toolbar. For more information about snapshots, refer to the [snapshot documentation]({{urlRoot}}/content/generating-a-snapshot).
+
+The GDK only generates schema for objects currently loaded into memory. This means if your project uses [sublevels](<https://docs.unrealengine.com/en-us/Engine/Levels/LevelsWindow>), you’ll need to load them in addition to your map, before generating schema.
+
+Schema [component IDs](https://docs.improbable.io/reference/13.5/shared/schema/reference#ids) persist between runs of the schema generator. This means that as long as you have generated schema for a particular map at least once, and have not modified any classes or maps since you last generated schema, then you only need to generate a snapshot for the current level before you start a SpatialOS deployment. 
+
+## Deleting schema
+
+Whenever you generate schema, the GDK checks the [`SchemaDatabase`(Unreal GDK Glossary)]({{urlRoot}}/content/glossary#schemadatabase) and all the relevant objects in your project and removes any references to missing objects from the `SchemaDatabase`.
```
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Good shout. I forgot to update this after settling on terminology. I think we went for SchemaDatabase when referring to it as a concept and `SchemaDatabase` when referring to the file itself. 

Fixed. (also fixed the earlier SchemaDatabase mention)

File: SpatialGDK/Documentation/content/schema.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+SpatialOS uses [schema (SpatialOS documentation)](https://docs.improbable.io/reference/13.6/shared/concepts/schema#schema) to generate APIs specific to the components in your game world. These APIs define how you can operate on entity components within SpatialOS. 
+
+## Generating schema
+
+To generate schema, select the **Schema** button in the [GDK Toolbar]({{urlRoot}}/content/toolbars#buttons). The GDK automatically iterates through objects with replicated properties to generate the required schema files and then updates the [`SchemaDatabase`(Unreal GDK Glossary)]({{urlRoot}}/content/glossary#schemadatabase).
+
+![Toolbar]({{assetRoot}}assets/screen-grabs/toolbar/schema-button.png)
+
+*Image: In the GDK toolbar in the Unreal Editor, select the* ***Schema*** *button.*
```
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Edited, ty.

File: SpatialGDK/Documentation/content/schema.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+As the GDK automatically generates all the schema you need, you do not have to write or edit schema manually when using the GDK.
+
+### When to generate schema
+
+You must generate schema:
+
+- When you add or change any replicated properties that you want to deploy to SpatialOS.
+- When you create a new map. This ensures that you generate schema for any unique objects on this map.
```
Comment (m-samiec) [inline SpatialGDK/Documentation/content/schema.md]: As snapshots don't contain any map specific information, this isn't required anymore.

File: SpatialGDK/Documentation/content/schema.md (hunk 6 above)
Comment (m-samiec) [inline SpatialGDK/Documentation/content/schema.md]: This is also not true now that generated snapshots don't contain any level information,

File: SpatialGDK/Documentation/content/schema.md [hunk 12]
```diff
@@ -0,0 +1,47 @@
+# Schema
+
+SpatialOS uses [schema (SpatialOS documentation)](https://docs.improbable.io/reference/13.6/shared/concepts/schema#schema) to generate APIs specific to the components in your game world. These APIs define how you can operate on entity components within SpatialOS. 
+
+## Generating schema
+
+To generate schema, select the **Schema** button in the [GDK Toolbar]({{urlRoot}}/content/toolbars#buttons). The GDK automatically iterates through objects with replicated properties to generate the required schema files and then updates the [`SchemaDatabase`(Unreal GDK Glossary)]({{urlRoot}}/content/glossary#schemadatabase).
```
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Per the discussion on the Gdoc, 'iterates' is the correct term in this context. 
> Iteration, in the context of computer programming, is a process wherein a set of instructions or structures are repeated in a sequence a specified number of times or until a condition is met.

File: SpatialGDK/Documentation/content/schema.md (hunk 11 above)
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Nuked as part of the snapshot changes. Thanks :D

File: SpatialGDK/Documentation/content/schema.md (hunk 6 above)
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Also nuked this bit.

File: SpatialGDK/Documentation/content/schema.md (hunk 11 above)
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Section now reads: 

```
You must generate schema when you add or change any replicated properties that you want to deploy to SpatialOS.
```

Is that correct?

File: SpatialGDK/Documentation/content/glossary.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/glossary.md]: Good suggestion. Edited.

[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/glossary.md]: Yeah UObjects is probably correct. Epic's documentation is inconsistent with this but I'm happy to go with UObjects in ours.

File: SpatialGDK/Documentation/content/schema.md [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+*Image: In the GDK toolbar in the Unreal Editor, select the* ***Schema*** *button.*
+
+As the GDK automatically generates all the schema you need, you do not have to write or edit schema manually when using the GDK.
+
+### When to generate schema
+
+You must generate schema:
+
+- When you add or change any replicated properties that you want to deploy to SpatialOS.
```
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: We don't have a page or a glossary entry for this. 
Could link here: https://docs.unrealengine.com/en-US/Gameplay/Networking/Actors/Properties
@m-samiec , thoughts?

File: SpatialGDK/Documentation/content/schema.md [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+![Toolbar]({{assetRoot}}assets/screen-grabs/toolbar/schema-button.png)
+
+*Image: In the GDK toolbar in the Unreal Editor, select the* ***Schema*** *button.*
+
+As the GDK automatically generates all the schema you need, you do not have to write or edit schema manually when using the GDK.
+
+### When to generate schema
+
+You must generate schema:
```
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Section removed as the schema generation requirements have changed slightly with the snapshot change.

File: SpatialGDK/Documentation/content/schema.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: We don't have a page for maps as they're a basic (albeit confusingly named) Unreal concept. Unreal also doesn't have a Map documentation page. They have a 'Level' documentation page, which is semi-relevant but also confusing. 

I've got a glossary entry in the works to try and make this less confusing.

File: SpatialGDK/Documentation/content/schema.md [hunk 15]
```diff
@@ -0,0 +1,47 @@
+# Schema
+
+SpatialOS uses [schema (SpatialOS documentation)](https://docs.improbable.io/reference/13.6/shared/concepts/schema#schema) to generate APIs specific to the components in your game world. These APIs define how you can operate on entity components within SpatialOS.
```
[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Per the canonology I think SpatialOS component is the correct term. @m-samiec might be able to comment here.

File: SpatialGDK/Documentation/content/schema.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (m-samiec) [reply SpatialGDK/Documentation/content/schema.md]: That link works 👍

File: SpatialGDK/Documentation/content/schema.md (hunk 11 above)
Comment (m-samiec) [reply SpatialGDK/Documentation/content/schema.md]: That is correct.

File: SpatialGDK/Documentation/content/schema.md (hunk 9 above)
Comment (m-samiec) [inline SpatialGDK/Documentation/content/schema.md]: Tiny nit, but schema generator operates on a per-class basis, not per-object.

Comment (m-samiec) [inline SpatialGDK/Documentation/content/schema.md]:
```suggestion
Whenever you generate schema, the GDK checks the [`SchemaDatabase`(Unreal GDK Glossary)]({{urlRoot}}/content/glossary#schemadatabase) and all the in-memory classes in your project and removes any classes referenced in the `SchemaDatabase` that no longer exist. 
```

Comment (m-samiec) [reply SpatialGDK/Documentation/content/schema.md]: Something like this, but with better words.

Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Nice suggestion! Edited.

Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Tiny nits are the best nits. Edited.

File: SpatialGDK/Documentation/content/schema.md [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+The GDK only generates schema for objects currently loaded into memory. This means if your project uses [sublevels](<https://docs.unrealengine.com/en-us/Engine/Levels/LevelsWindow>), you’ll need to load them in addition to your map, before generating schema.
+
+Schema [component IDs](https://docs.improbable.io/reference/13.5/shared/schema/reference#ids) persist between runs of the schema generator. This means that as long as you have generated schema for a particular map at least once, and have not modified any classes or maps since you last generated schema, then you only need to generate a snapshot for the current level before you start a SpatialOS deployment. 
+
+## Deleting schema
+
+Whenever you generate schema, the GDK checks the [`SchemaDatabase`(Unreal GDK Glossary)]({{urlRoot}}/content/glossary#schemadatabase) and all the relevant objects in your project and removes any references to missing objects from the `SchemaDatabase`. 
+
+If you delete any objects in your project, the GDK removes them from the `SchemaDatabase` the next time you generate schema.
```
Comment (m-samiec) [inline SpatialGDK/Documentation/content/schema.md]: Also here, any reference to `object` from here on should probably be swapped with `class`

Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Done. Thanks!

File: SpatialGDK/Documentation/content/schema.md (hunk 9 above)
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: (They are fine words)

File: SpatialGDK/Documentation/content/schema.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (m-samiec) [reply SpatialGDK/Documentation/content/schema.md]: Yes, just component is fine.

[the remark this answers is not in the document]
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: Is it more correct to say:
>to generate APIs specific to the components in your project. These APIs define how you can operate on entity components within SpatialOS.
?

[the remark this answers is not in the document]
Comment (m-samiec) [reply SpatialGDK/Documentation/content/schema.md]: Sounds good.

File: SpatialGDK/Documentation/content/generating-a-snapshot.md [hunk 17]
```diff
@@ -1,22 +1,27 @@
...
-Startup Actors are [stably named replicated Actors (Unreal documentation)](https://docs.unrealengine.com/en-US/Gameplay/Networking/Actors/Properties/ObjectReferences) that you have placed in a Level. The startup Actors need to be listed in the snapshot so that SpatialOS spawns an entity only once for each Actor. If the startup Actors are not in the snapshot, you may get multiple entities for each Actor when you launch multiple [server-workers]({{urlRoot}}/content/glossary#workers). 
+_Image: The GDK Toolbar._
+
+This creates a snapshot called `default.snapshot` which you can find in `spatial\snapshots`.
+
+If you want your snapshots to be exported to a different path you can specify the output path and file name of the snapshot using the [GDK toolbar settings]({{urlRoot}}/content/toolbars.md).
+
+## What’s listed in snapshots
+
+The GDK snapshots contain three kinds of [SpatialOS entities]({{urlRoot}}/content/glossary#spatialos-entity):
```
Comment (m-samiec) [inline SpatialGDK/Documentation/content/generating-a-snapshot.md]:
```suggestion
The GDK snapshots contain two kinds of [SpatialOS entities]({{urlRoot}}/content/glossary#spatialos-entity): 
```

File: SpatialGDK/Documentation/content/schema.md [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+
+### When to generate schema
+
+You must generate schema when you add or change any [replicated properties (Unreal documentation)](https://docs.unrealengine.com/en-US/Gameplay/Networking/Actors/Properties) that you want to deploy to SpatialOS.
+
+The GDK only generates schema for classes currently loaded into memory. This means if your project uses [sublevels](<https://docs.unrealengine.com/en-us/Engine/Levels/LevelsWindow>), you’ll need to load them in addition to your map, before generating schema.
+
+## Deleting schema
+
+Whenever you generate schema, the GDK checks the [SchemaDatabase]({{urlRoot}}/content/glossary#schemadatabase) and all of your project classes currently loaded into memory, and removes any classes referenced in the `SchemaDatabase` that no longer exist.
```
Comment (m-samiec) [inline SpatialGDK/Documentation/content/schema.md]: This sentence and the next kinda say the same thing. Essentially - "When generating schema, the GDK verifies that anything previously saved in the SchemaDatabase still exists. If the class has been deleted, it's also removed from the SchemaDatabase."

File: SpatialGDK/Documentation/content/schema.md [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+
+# Schema and source control 
+
+### Checking out the `SchemaDatabase`
+
+**Note:** If you are using the built-in [Unreal source control system](https://docs.unrealengine.com/en-US/Engine/UI/SourceControl) Unreal locks this file on checkout, meaning other users are unable to write to it. To prevent this, mark the `SchemaDatabase` file as writable locally on each machine, and only check out the file when you are ready to commit any changes made to it. 
+
+## Schema for unique local classs
+
+Whenever you generate schema, the GDK checks the [SchemaDatabase]({{urlRoot}}/content/glossary#schemadatabase) and all the in-memory classes in your project and removes any classes referenced in the `SchemaDatabase` that no longer exist.
```
Comment (m-samiec) [inline SpatialGDK/Documentation/content/schema.md]: This sentence contains duplicate info from above.

File: SpatialGDK/Documentation/content/schema.md (hunk 18 above)
Comment (NotRobSheridan) [reply SpatialGDK/Documentation/content/schema.md]: >New text
When you generate schema, the GDK verifies that any classes referenced in the [SchemaDatabase]({{urlRoot}}/content/glossary#schemadatabase) still exist. If you delete a class the GDK removes it from the SchemaDatabase the next time you generate schema.

**Pull request merged**